netbox/core/plugins.py

```diff
@@ -0,0 +1,195 @@
+import datetime
+import importlib
+import importlib.util
+from dataclasses import dataclass, field
+from typing import Optional
+
+import requests
+from django.conf import settings
+from django.core.cache import cache
+from django.utils.translation import gettext_lazy as _
+
+from netbox.plugins import PluginConfig
+from utilities.datetime import datetime_from_timestamp
+
+
+@dataclass
+class PluginAuthor:
+    """
+    Identifying information for the author of a plugin.
+    """
+    name: str
+    org_id: str = ''
+    url: str = ''
+
+
+@dataclass
+class PluginVersion:
+    """
+    Details for a specific versioned release of a plugin.
+    """
+    date: datetime.datetime = None
+    version: str = ''
+    netbox_min_version: str = ''
+    netbox_max_version: str = ''
+    has_model: bool = False
+    is_certified: bool = False
+    is_feature: bool = False
+    is_integration: bool = False
+    is_netboxlabs_supported: bool = False
+
+
+@dataclass
+class Plugin:
+    """
+    The representation of a NetBox plugin in the catalog API.
+    """
+    id: str = ''
+    status: str = ''
+    title_short: str = ''
+    title_long: str = ''
+    tag_line: str = ''
+    description_short: str = ''
+    slug: str = ''
+    author: Optional[PluginAuthor] = None
+    created_at: datetime.datetime = None
+    updated_at: datetime.datetime = None
+    license_type: str = ''
+    homepage_url: str = ''
+    package_name_pypi: str = ''
+    config_name: str = ''
+    is_certified: bool = False
+    release_latest: PluginVersion = field(default_factory=PluginVersion)
+    release_recent_history: list[PluginVersion] = field(default_factory=list)
+    is_local: bool = False  # extra field for locally installed plugins
+    is_installed: bool = False
+    installed_version: str = ''
+
+
+def get_local_plugins():
+    """
+    Return a dictionary of all locally-installed plugins, mapped by name.
+    """
+    plugins = {}
+    for plugin_name in settings.PLUGINS:
+        plugin = importlib.import_module(plugin_name)
+        plugin_config: PluginConfig = plugin.config
+
+        plugins[plugin_config.name] = Plugin(
+            slug=plugin_config.name,
+            title_short=plugin_config.verbose_name,
+            tag_line=plugin_config.description,
+            description_short=plugin_config.description,
+            is_local=True,
+            is_installed=True,
+            installed_version=plugin_config.version,
+        )
+
+    return plugins
+
+
+def get_catalog_plugins():
+    """
+    Return a dictionary of all entries in the plugins catalog, mapped by name.
+    """
+    session = requests.Session()
```

Comment:
For catalog API requests can we set a custom `User-Agent` header of a form similar to:

`NetBox/{settings.RELEASE.version} (Python {platform.python_version()}; {platform.system()} {platform.machine()}) {settings.RELEASE.edition}`

Example value:

`NetBox/v4.1.0 (Python 3.10.12; Linux x86_64) Community`